### PR TITLE
R21 sc 205 add supercontract txn

### PR DESCRIPTION
**Pull request by cyrillicw**

### CODE REVIEW

File: sdk/supercontract_v2_test.go [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+              0
+            ]
+          }
+        ]
+      }
+    ],
+    "releasedTransactions": [
+      {
+        "releasedTransactionHash": "0100000000000000000000000000000000000000000000000000000000000000"
+      }
```
Comment (jenyasd209) [inline sdk/supercontract_v2_test.go:115]: see comment about `releasedTransactionHashDto`

File: sdk/supercontract_v2_transaction.go [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+	actualArgumentsSize 	uint16,
+	executionCallPayment 	Amount,
+	downloadCallPayment 	Amount,
+	servicePaymentsCount 	uint8,
+	fileName 				string,
+	functionName 			string,
+	actualArguments 		string,
+	servicePayments 		[]*MosaicId,
+	networkType 			NetworkType,
+) (*ManualCallTransaction, error) {
```
Comment (jenyasd209) [inline sdk/supercontract_v2_transaction.go:133]: see comment about redundant fields.

Also, looks like this file and some other have problems with formatting. Fix it with `go fmt` tool,  pleas.

File: sdk/transaction_model.go [hunk 3]
```diff
@@ -2903,8 +2906,11 @@ const (
 	ReplicatorOffboarding          EntityType = 0x4762
 	CreateLiquidityProvider        EntityType = 0x4169
 	ManualRateChange               EntityType = 0x4269
-	PlaceSdaExchangeOffer     EntityType = 0x416A
-	RemoveSdaExchangeOffer    EntityType = 0x426A
+	PlaceSdaExchangeOffer     	   EntityType = 0x416A
+	RemoveSdaExchangeOffer    	   EntityType = 0x426A
+	DeployContract				   EntityType = 0x416E
+	ManualCall					   EntityType = 0x426E
+	AutomaticExecutionsPayment	   EntityType = 0x436E
```
Comment (jenyasd209) [inline sdk/transaction_model.go:2913]: formatting

Comment (jenyasd209) [reply sdk/transaction_model.go:2913]: and add types and versions for `Successful and Unsuccessful EndBatchExecutionTransaction`

File: sdk/supercontract_v2_transaction.go [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+	}
+
+	return &tx, nil
+}
+
+func (tx *DeployContractTransaction) GetAbstractTransaction() *AbstractTransaction {
+	return &tx.AbstractTransaction
+}
+
+func (tx *DeployContractTransaction) Bytes() ([]byte, error) {
```
Comment (jenyasd209) [inline sdk/supercontract_v2_transaction.go:341]: `DeployContractTransaction` and `ManualCallTransaction` have a lot of common fields. Maybe it makes sense to move them to a separate func, where these fields will be converted to vectors.

**Review by jenyasd209 — CHANGES_REQUESTED** [review]
Also:
- `sdk.go` lacks new sc_V2 transactions;
- the`MapTransaction` funds (`transaction_model.go`) doesn't have cases for new transactions;
- fix formatting.

**Comment by jenyasd209** [comment]
@jackshenn

File: sdk/supercontract_v2_transaction.go (hunk 4 above)
Comment (jackshenn) [reply sdk/supercontract_v2_transaction.go:329]: https://github.com/proximax-storage/go-xpx-chain-sdk/commit/9fb76208c966da7be86d54c92222ac996ed3a42c
I have made some changes on this issue, is the changes ok?

File: sdk/supercontract_v2_internal.go [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+		info, err := dto.toStruct(networkType)
+		if err != nil {
+			return nil, err
+		}
+
+		releasedTransactions = append(releasedTransactions, info)
+	}
+
+	return releasedTransactions, nil
+}
```
[the remark this answers is not in the document]
Comment (jackshenn) [reply sdk/supercontract_v2_internal.go]: https://github.com/proximax-storage/go-xpx-chain-sdk/commit/1a52d84af3945dfe4d5a52358fe5782958873991
I have made some changes on this issue, is the changes ok?

[the remark this answers is not in the document]
Comment (jenyasd209) [reply sdk/supercontract_v2_internal.go]: Not really. In `superContractV2DTO` it's `string array`, but in `SuperContractV2` struct it should be `array of Hashes`

File: sdk/transaction_model.go [hunk 6]
```diff
@@ -3282,6 +3302,12 @@ func MapTransaction(b *bytes.Buffer, generationHash *Hash) (Transaction, error)
 		dto = &placeSdaOfferTransactionDTO{}
 	case RemoveSdaExchangeOffer:
 		dto = &removeSdaExchangeOfferTransactionDTO{}
+	case AutomaticExecutionsPayment:
+		dto = &automaticExecutionsPaymentTransactionDTO{}
+	case ManualCall:
+		dto = &manualCallTransactionDTO{}
+	case DeployContract:
+		dto = &deployContractTransactionDTO{}
```
Comment (jenyasd209) [inline sdk/transaction_model.go:3310]: also, add `Unsuccessful and SuccessfulEndBatchExecution`

File: sdk/sdk.go [hunk 7]
```diff
@@ -1031,6 +1033,95 @@ func (c *Client) NewManualRateChangeTransaction(
...
+}
+
+func (c *Client) NewManualCallTransaction(
+	deadline *Deadline,
+	contractKey *PublicAccount,
+	executionCallPayment Amount,
+	downloadCallPayment Amount,
+	fileName string,
+	functionName string,
+	actualArguments string,
```
Comment (jenyasd209) [inline sdk/sdk.go:1062]: `actualArguments` is `string` or `byte array`? Check, please.

Comment (jackshenn) [reply sdk/sdk.go:1062]: https://github.com/proximax-storage/go-xpx-chain-sdk/commit/d65df18c9b8d3f15f43f19350c10a2291a0271f4

File: sdk/transaction_model.go (hunk 6 above)
Comment (jackshenn) [reply sdk/transaction_model.go:3316]: https://github.com/proximax-storage/go-xpx-chain-sdk/commit/622604afaa00afc762c77f69247af97dffa8b036

File: sdk/supercontract_v2_internal.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jackshenn) [reply sdk/supercontract_v2_internal.go]: https://github.com/proximax-storage/go-xpx-chain-sdk/commit/7a3a286f592bc5323af4a8c735e32a8615bbcc19

**Review by jenyasd209 — APPROVED** [review]
@jackshenn  Looks good. If we missed something, will fix during acceptance testing.

**Pull request merged**